### DISCUSSION TITLE
DDR4 SDRAM - Understanding the Basics

### DISCUSSION THREAD

**Discussion by subbdue** [subbdue/systemverilog.io#2] [Website Comments]
Please use this Discussions thread for questions and comments on article https://www.systemverilog.io/ddr4-basics

**Comment by jredders** [comment]
The Tip right above DRAM Page Size has a typo:
So, you can buy a 4Gb cabinet which can hold A5 size paper(x4) or A4 size paper (x8) or A~~5~~3 size paper (x16).

And in the section under DRAM Size Calculation you're missing an s in two places:
Total number of row**s** = 2^16 = 64K
Total number of row**s** = 2^15 = 32K

**Reply by subbdue** [reply]
Thanks @jredders this has been fixed now!

**Comment by velidurmuscan** [comment]
There is a typo:

Original Text:
Number of Row Address bits: A0-A154 = 15 bits

Suggestion:
Number of Row Address bits: A0-A14 = 15 bits

**Reply by subbdue** [reply]
Thanks @velidurmuscan I've fixed this.

**Comment by lvordu** [comment]
There is a typo at "Partial Command Truth-Table": it should be "We_n/A14", not A4 I guess.

**Reply by subbdue** [reply]
@lvordu, Thanks I've fixed this.

**Comment by lcwilch** [comment]
Is there a mistake in "DRAM Size Calculation" section, the first example "/* 4Gb x4 Device */"?

/* 4Gb x4 Device */
Number of Row Address bits: **A0-A15 = 16 bits** should be **A0-A17** per fig 5 (JESD79-4B spec) 4Gb x4 device?
    then **Total number of rows = 2^16 = 64K** should this instead be **Total number of rows = 2^18 = 256K**? 
Number of Column Address bits: A0-A9 = 10 bits
    Number of columns per row = 1K
Width of each column = 4 bits
Number of Bank Groups = 4
Number of Banks = 4

Total DRAM Capacity = 
    Num.Rows x 
    Num.Columns x Width.of.Column x 
    Num.BankGroups x Num.Banks

Total DRAM Capacity = 
    **64K x 1K x 4 x 4 x 4 = 4Gb** would go to **256K x 1K x 4 x 4 x 4 = 16Gb**?

Looking back at this, perhaps the first comment was intended to be "/* 1Gb x4 Device *\/" ?  Or maybe it should be "/* 4Gb total from 1Gb x4 device*\/"? Or maybe I am misunderstanding.

Thank you so much for the write-up! It is extremely helpful and well written.

**Reply by subbdue** [reply]
Hi @lcwilch ... Thanks for your question and apologies for this late reply. 

It's not a mistake. I should probably go back and clarify this in my article. When memory manufacturers specify the DRAM configuration this is how they usually do it : `Total Density` `Width`

So in my example `4Gb` is the total density and `x4` is its width configuration (Because a 4Gb is also available in x8 and x16 widths). So when a device is specified as a 4Gb x4, you will need to look up the **4Gb Addressing Table** and find the **x4** width within that. Attaching an annotated picture below.

<img width="648" alt="Screenshot 2024-06-08 at 5 38 28 AM" src="https://github.com/subbdue/systemverilog.io/assets/2192400/25ef4099-d597-4516-87eb-4d58108c6dad">

This is a screenshot from the Micron website for a 4Gb part:
<img width="302" alt="image" src="https://github.com/subbdue/systemverilog.io/assets/2192400/14318c2e-4176-46bd-bd92-60257321cb44">
https://www.micron.com/products/memory/dram-components/ddr4-sdram/part-catalog/part-detail/mt40a256m16ly-062e-f

Hope this clarifies your question. Please do write back if it still isn't clear, or if you disagree with something.

**Comment by Aaronyap2002** [comment]
```The PHY then does all the lower level signaling and drives the physical interface to the DRAM.```
```Think of the controller as the brains and the PHY as the brawns.```
I don't understand the purpose of the PHY. Why do we need the PHY in the first place? Isn't the controller suppose to do the signaling?

**Reply by subbdue** [reply]
@Aaronyap2002 External interfaces, such as, PCIe, DDR, SerDes ethernet require a PHY to drive and receive logic 1s and 0s to/from the PCB traces. 

On the transmitter side, the PHY is primarily concerned with making sure 1s and 0s can be reliably driven out, and on the receiving side its main job is to reliably recover 1s and 0s that arrive into the pins (or balls) of the SoC. To enable this, it usually has Analog Blocks and uses Equalization techniques such as CTLE, FFE, DFE, etc. 

The Controllers on the other hand are purely digital. Their main purpose is to ensure that the interface protocol is followed. In case of DDR, the controller ensures that timing parameters are not violated -- for example, making sure during writes the WriteData is launched WL (Write Latency) clocks after the WriteCommand is issued; or when switching from a Write to a Read, there is a tWTR (Write-to-Read turnaround time) gap between them.

**Comment by TobyHarry** [comment]
Just started learning about DDR. Can anyone explain if the so called "Page break" and "bank collision" the same? if not, what is the difference? Are they both causing performance penalty? 
Thanks

**Comment by nightseas** [comment]
Should it be two 8Gb x 4 chips to get same capacity of 8Gb x 8?
![image](https://github.com/user-attachments/assets/75a5756a-6137-40b8-8cce-d4d069133db7)

**Comment by ConvolutedDog** [comment]
Hi Subbdue,

I have a questions: For an x8 DDR4 device with 8 memory arrays per bank, can it support burst length of 4 (BC4) mode?

Tanks!
